csi/controller_server.go

```diff
@@ -268,8 +268,13 @@ func (cs *ControllerServer) ControllerPublishVolume(ctx context.Context, req *cs
 		return nil, status.Errorf(codes.Aborted, "The volume %s is %s", req.GetVolumeId(), existVol.State)
 	}
 
+	// Check volume frontend settings
+	if existVol.DisableFrontend || existVol.Frontend != string(types.VolumeFrontendBlockDev) {
+		return nil, status.Errorf(codes.InvalidArgument, "ControllerPublishVolume: there is no block device frontend for volume %s", req.GetVolumeId())
+	}
+
 	// the volume is already attached make sure it's to the same node as this
-	if existVol.State == string(types.VolumeStateAttached) {
+	if existVol.State == string(types.VolumeStateAttached) && existVol.Controllers[0].Endpoint != "" {
 		if !existVol.Ready || len(existVol.Controllers) == 0 {
```

Comment:
add || existVol.Controllers[0].Endpoint == "" to this volume ready for workload check.
Since that will return aborted, which is the code for operation in progress.

Might even look into only having the volume be ready once an endpoint is available, but I would have to look more at the code to make sure that this would be the correct behavior. cc @yasker